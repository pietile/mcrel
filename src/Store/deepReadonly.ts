// Deep readonlyy object
// See https://github.com/microsoft/TypeScript/pull/21316s
export type DeepReadonlyObject<T> = {
  readonly [P in keyof T]: T[P] extends object
    ? T[P] extends (...args: any[]) => any
      ? T[P]
      : DeepReadonly<T[P]>
    : T[P];
};

interface DeepReadonlyArray<T> extends ReadonlyArray<DeepReadonly<T>> {}
interface DeepReadonlyMap<K, V> extends ReadonlyMap<DeepReadonly<K>, DeepReadonly<V>> {}
interface DeepReadonlySet<V> extends ReadonlySet<DeepReadonly<V>> {}

type DeepReadonly<T> = T extends ReadonlyArray<infer V>
  ? DeepReadonlyArray<V>
  : T extends ReadonlyMap<infer K, infer V>
  ? DeepReadonlyMap<K, V>
  : T extends ReadonlySet<infer V>
  ? DeepReadonlySet<V>
  : DeepReadonlyObject<T>;
