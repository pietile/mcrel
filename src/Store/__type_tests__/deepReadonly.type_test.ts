import { DeepReadonlyObject } from '../deepReadonly';

interface SubState {
  a: string;
  subObj: {
    b: number;
  };
  array: string[];
}

interface State {
  a: number;
  subObj: {
    b: SubState;
  };
  array: SubState[];
  map: Map<SubState, SubState>;
  set: Set<SubState>;
  readonlysubObj: {
    readonly c: SubState;
  };
  readonlyArray: ReadonlyArray<SubState>;
  readonlyMap: ReadonlyMap<SubState, SubState>;
  readonlySet: ReadonlySet<SubState>;
}

type ReadonlyState = DeepReadonlyObject<State>;

const sub: SubState = {
  a: 'a',
  subObj: {
    b: 0,
  },
  array: ['a', 'b'],
};

// @dts-jest:pass:snap
const obj: ReadonlyState = {
  a: 0,
  subObj: {
    b: sub,
  },
  array: [sub],
  map: new Map([[sub, sub]]),
  set: new Set([sub]),
  readonlysubObj: {
    c: sub,
  },
  readonlyArray: [sub],
  readonlyMap: new Map([[sub, sub]]),
  readonlySet: new Set([sub]),
};

// @dts-jest:fail:snap
obj.a = 1;

// @dts-jest:fail:snap
obj.subObj.b = sub;
// @dts-jest:fail:snap
obj.subObj.b.a = 1;
// @dts-jest:fail:snap
obj.subObj.b.subObj.b = 1;
// @dts-jest:fail:snap
obj.subObj.b.array.pop();

{
  const value = obj.map.get(sub);
  if (value) {
    // @dts-jest:fail:snap
    value.a = 1;
  }
}

obj.set.forEach((value) => {
  // @dts-jest:fail:snap
  value.a = 1;
});

// @dts-jest:fail:snap
obj.readonlysubObj.c = sub;
// @dts-jest:fail:snap
obj.readonlysubObj.c.a = 1;
// @dts-jest:fail:snap
obj.readonlysubObj.c.subObj.b = 1;
// @dts-jest:fail:snap
obj.readonlysubObj.c.array.pop();

{
  const value = obj.readonlyMap.get(sub);
  if (value) {
    // @dts-jest:fail:snap
    value.a = 1;
  }
}

obj.readonlySet.forEach((value) => {
  // @dts-jest:fail:snap
  value.a = 1;
});
