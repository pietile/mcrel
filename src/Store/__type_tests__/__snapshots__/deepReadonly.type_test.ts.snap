// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`const obj: ReadonlyState = {
    a: 0,
    subObj: {
      b: sub,
    },
    array: [sub],
    map: new Map([[sub, sub]]),
    set: new Set([sub]),
    readonlysubObj: {
      c: sub,
    },
    readonlyArray: [sub],
    readonlyMap: new Map([[sub, sub]]),
    readonlySet: new Set([sub]),
  } (type) should match snapshot 1`] = `"any"`;

exports[`obj.a = 1 (type) should match snapshot 1`] = `"Cannot assign to 'a' because it is a read-only property."`;

exports[`obj.readonlysubObj.c = sub (type) should match snapshot 1`] = `"Cannot assign to 'c' because it is a read-only property."`;

exports[`obj.readonlysubObj.c.a = 1 (type) should match snapshot 1`] = `"Cannot assign to 'a' because it is a read-only property."`;

exports[`obj.readonlysubObj.c.array.pop() (type) should match snapshot 1`] = `"Property 'pop' does not exist on type 'DeepReadonlyArray<string>'."`;

exports[`obj.readonlysubObj.c.subObj.b = 1 (type) should match snapshot 1`] = `"Cannot assign to 'b' because it is a read-only property."`;

exports[`obj.subObj.b = sub (type) should match snapshot 1`] = `"Cannot assign to 'b' because it is a read-only property."`;

exports[`obj.subObj.b.a = 1 (type) should match snapshot 1`] = `"Cannot assign to 'a' because it is a read-only property."`;

exports[`obj.subObj.b.array.pop() (type) should match snapshot 1`] = `"Property 'pop' does not exist on type 'DeepReadonlyArray<string>'."`;

exports[`obj.subObj.b.subObj.b = 1 (type) should match snapshot 1`] = `"Cannot assign to 'b' because it is a read-only property."`;

exports[`value.a = 1 (type) should match snapshot 1`] = `"Cannot assign to 'a' because it is a read-only property."`;

exports[`value.a = 1 (type) should match snapshot 2`] = `"Cannot assign to 'a' because it is a read-only property."`;

exports[`value.a = 1 (type) should match snapshot 3`] = `"Cannot assign to 'a' because it is a read-only property."`;

exports[`value.a = 1 (type) should match snapshot 4`] = `"Cannot assign to 'a' because it is a read-only property."`;
